Add PM Target Cycling to Chat History plugin

TheStonedTurtle opened the pull request:
Pressing tab again while typing a private message will change the target recipient by cycling through the list of recipients

![2018-12-08_10-29-38](https://user-images.githubusercontent.com/29030969/49689323-5f14a780-fad4-11e8-8b61-4d4dee5c6bbf.gif)

Closes #7885

TheStonedTurtle commented:
PR has been updated to invoke the jagex script when changing targets and to grab the current target name via the `VarClientStrChanged` event

Abextm commented on runelite-api/src/main/java/net/runelite/api/ScriptID.java:
```diff
@@ -45,6 +45,16 @@
+	 *
+	 * Jagex refers to this script as {@code meslayer_mode6}
+	 * <ul>
+	 * <li> String Player to send private message too</li>
```
```suggestion
	 * <li> String Player to send private message to </li>
```

Abextm commented on runelite-client/src/main/java/net/runelite/client/plugins/pmcycle/PMCyclePlugin.java:
```diff
@@ -0,0 +1,142 @@
+
+			if (client.getVar(VarClientInt.INPUT_TYPE) == InputType.PRIVATE_MESSAGE.getType())
+			{
+				String newFriend = friends.lower(target);
```
Anything touching `friends` should happen on the clientThread

TheStonedTurtle commented:
The requested changes have been made also using a LinkedHashSet instead of a TreeSet

TheStonedTurtle commented:
I've changed this to loop over the LinkedHashSet instead of converting it to an ArrayList.

TheStonedTurtle commented:
Rebased PR & Cleaned up commit history

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/pmcycle/PMCyclePlugin.java:
```diff
@@ -0,0 +1,155 @@
+@Slf4j
+@PluginDescriptor(
+	name = "Private Message Cycle",
+	description = "When sending a private message pressing the tab button again will cycle through your previous recipients"
```
This description should be shorter and make more sense

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/pmcycle/PMCyclePlugin.java:
```diff
@@ -0,0 +1,155 @@
+	@Inject
+	private KeyManager keyManager;
+
+	private final LinkedHashSet<String> friends = new LinkedHashSet<>();
```
This should probably just be a fixed size `Deque`. It certainly should not grow unbounded like you have here.

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/pmcycle/PMCyclePlugin.java:
```diff
@@ -0,0 +1,155 @@
+	{
+		if (c.getIndex() == VarClientStr.PRIVATE_MESSAGE_TARGET.getIndex())
+		{
+			target = client.getVar(VarClientStr.PRIVATE_MESSAGE_TARGET);
```
Why can't this just be fetched in `keyPressed`?

Kamielvf commented:
This should go into the `Chat History` plugin IMO.

TheStonedTurtle commented:
Implemented Review changes and now conserves message content when changing targets

Also rebased to use `ChatMessage` since `SetMessage` was removed.

ShaunDreclin commented:
Would this let you reply to the last pm sent to you even if you're not currently sending one? Being able to just hit tab to reply to people would be awesome

TheStonedTurtle commented:
> Being able to just hit tab to reply to people would be awesome

This is a feature jagex already has implemented. Have you gone this whole time right-click responding to PMs?

ShaunDreclin commented:
No of course not

🙄

Adam- commented:
I agree this would make the most sense as part of the chathistory plugin.

TheStonedTurtle commented:
I'll add this to the ChatHistory plugin then. I assume this functionality should have a config option?

Hydrox6 commented:
If it's going to be added there, can a toggle be added to not have the chat history itself?

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/pmcycle/PMCyclePlugin.java:
```diff
@@ -0,0 +1,149 @@
+	@Inject
+	private KeyManager keyManager;
+
+	private final EvictingQueue<String> friends = EvictingQueue.create(100);
```
this should be ~5

TheStonedTurtle commented:
Updated PR to add the PM Cycling functionality to the Chat History plugin

Gif after update:
https://gfycat.com/MildCompetentGourami

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryConfig.java:
```diff
@@ -0,0 +1,55 @@
+	@ConfigItem(
+		keyName = "pmTargetCycling",
+		name = "PM Target Cycling",
+		description = "Retains chat history when logging in/out or world hopping",
```
This needs an updated description

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryPlugin.java:
```diff
@@ -57,22 +73,53 @@
 	);
 
+	private static final int CYCLE_HOTKEY = KeyEvent.VK_TAB;
+	private static final Set<ChatMessageType> CYCLE_TYPES = Sets.newHashSet(
```
`ImmutableSet`

TheStonedTurtle replied:
I'll update the one right above this as well then.

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryPlugin.java:
```diff
@@ -92,7 +139,7 @@ public void onChatMessage(ChatMessage chatMessage)
 		}
 
-		if (ALLOWED_HISTORY.contains(chatMessage.getType()))
+		if (config.retainChatHistory() && ALLOWED_HISTORY.contains(chatMessage.getType()))
```
This config opt should be on the replay.

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryPlugin.java:
```diff
@@ -143,4 +199,64 @@ private static String nbsp(final String message)
+	@Override
+	public void keyPressed(KeyEvent e)
+	{
+		if (!config.pmTargetCycling() || e.getKeyCode() != CYCLE_HOTKEY)
```
Check keycode first.

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryPlugin.java:
```diff
@@ -143,4 +199,64 @@ private static String nbsp(final String message)
+		clientThread.invoke(() ->
+		{
+			final String currentTarget = client.getVar(VarClientStr.PRIVATE_MESSAGE_TARGET);
+			if (currentTarget == null || friends.size() == 0)
```
This should be `isEmpty` - and move it to `findPreviousFriend`

TheStonedTurtle commented:
Requested review changes have been implemented.

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryPlugin.java:
```diff
@@ -57,30 +72,61 @@
 		// Start sending old messages right after the welcome message, as that is most reliable source
 		// of information that chat history was reset
-		if (chatMessage.getMessage().equals(WELCOME_MESSAGE))
+		if (config.retainChatHistory() && chatMessage.getMessage().equals(WELCOME_MESSAGE))
```
This is a behavior change and makes it start retaining the welcome message. Maybe don't do this.

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryPlugin.java:
```diff
@@ -108,6 +154,15 @@ public void onChatMessage(ChatMessage chatMessage)
+		if (CYCLE_TYPES.contains(chatMessage.getType()))
+		{
+			final String name = Text.removeTags(chatMessage.getName());
+			// Remove to ensure uniqueness & it's place in history
```
`its`

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryPlugin.java:
```diff
@@ -108,6 +154,15 @@ public void onChatMessage(ChatMessage chatMessage)
+			// Remove to ensure uniqueness & it's place in history
+			friends.remove(name);
+			friends.add(name);
+
```
Remove this line.

TheStonedTurtle commented:
Fixed commits to address review comments

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/chathistory/ChatHistoryPlugin.java:
```diff
@@ -108,6 +159,14 @@ public void onChatMessage(ChatMessage chatMessage)
 			}
 		}
+
+		if (CYCLE_TYPES.contains(chatMessage.getType()))
```
This was just a switch in the other PR which seems like it is probably cleaner?

TheStonedTurtle replied:
switched to a switch statement
